packages/terra-menu/src/Menu.jsx

```diff
@@ -0,0 +1,215 @@
+    this.getInitialState = this.getInitialState.bind(this);
+    this.push = this.push.bind(this);
+    this.pop = this.pop.bind(this);
+    this.state = this.getInitialState();
+  }
+
+  getInitialState() {
+    const initialMenu = (
+      <MenuContent
+        key="MenuPage-0"
+        onRequestNext={this.push}
+        onRequestBack={this.pop}
+        onRequestClose={this.props.onRequestClose}
+        index={0}
+      >
+        {this.props.children}
+      </MenuContent>
+    );
+
+    return {
+      stack: [initialMenu],
+    };
+  }
+
+  componentWillReceiveProps(nextProps) {
+    if (this.props.isOpen && !nextProps.isOpen) {
+      this.setState(this.getInitialState());
+    }
+  }
+
+  getContentHeight() {
+    let itemCount = 0;
+    let dividerCount = 0;
+
+    React.Children.forEach(this.props.children, (child) => {
+      if (child.props.children && child.props.children.length > 0) {
+        itemCount += child.props.children.length;
+      } else if (child.type === <MenuDivider />.type) {
+        dividerCount += 1;
+      } else {
+        itemCount += 1;
+      }
+    });
+
+    return (itemCount * MENU_ITEM_HEIGHT) + (dividerCount * MENU_DIVIDER_HEIGHT) + MENU_PADDING_TOP + MENU_PADDING_BOTTOM;
+  }
+
+  pop() {
+    this.setState((prevState) => {
+      prevState.stack.pop();
+      return { stack: prevState.stack };
+    });
+  }
+
+  push(item) {
+    const index = this.state.stack.length;
+    const content = (
+      <MenuContent
+        key={`MenuPage-${index}`}
+        title={item.props.text}
+        index={index}
+        onRequestBack={this.pop}
+        onRequestClose={this.props.onRequestClose}
+        onRequestNext={this.push}
+      >
+        {item.props.subMenuItems}
+      </MenuContent>
+    );
+
+    this.setState((prevState) => {
+      prevState.stack.push(content);
+      return { stack: prevState.stack };
+    });
+  }
+
+  render() {
+    const {
+      boundingRef,
+      classNameArrow,
+      classNameContent,
+      classNameOverlay,
+      onRequestClose,
+      isOpen,
+      children,
+      targetRef,
+      isArrowDisplayed,
+      contentWidth,
+      ...customProps
+    } = this.props;
+    const attributes = Object.assign({}, customProps);
+    const contentHeight = this.getContentHeight();
+    const popupHeight = MenuUtils.getPopupHeight(contentHeight);
+    const boundingFrame = this.props.boundingRef ? this.props.boundingRef() : undefined;
+    const isFullScreen = MenuUtils.isFullScreen(boundingFrame, popupHeight, MenuWidths[this.props.contentWidth]);
+
+    const isSubMenu = this.state.stack.length > 1;
+    const contentClass = cx([
+      { submenu: isSubMenu },
+      { 'main-menu': !isSubMenu },
+      { fullscreen: isFullScreen },
+      classNameContent,
+    ]);
+
+    const arrowClass = cx([
+      'arrow',
+      { submenu: isSubMenu },
+      classNameArrow,
+    ]);
+
+    return (
+      <Popup
+        {...attributes}
```

Comment:
this could be spread directly with {...customProps} with reassignment to attributes